### PR TITLE
Server Access Integration

### PR DESCRIPTION
**Pull request by xacrimon**
This PR contains changes and primitives needed for integration with Server Access.

Paired with https://github.com/gravitational/teleport/pull/18020

Changes:
- [x] Update code translation to generate correct calls for xor & friends
- [x] Add a general `Resource` primitive for describing read/list access to resources.
- [x] Add an example policy allowing a user to list all nodes and join as root.

### CODE REVIEW

File: predicate/solver/teleport.py [hunk 1]
```diff
@@ -172,6 +186,15 @@ class User:
     # name is username
     name = ast.String("user.name")
 
+    # policies is a list of access policies assigned to the user
+    polices = ast.StringList("user.policies")
+
+    # roles is a list of roles assigned to the user
+    roles = ast.StringList("user.roles")
+
+    # policies is a list of logins/principals assigned to the user
```
Comment (klizhentas) [inline predicate/solver/teleport.py:195]: `ssh_logins` is a list of logins/principals (typo)

File: predicate/solver/teleport.py [hunk 2]
```diff
@@ -172,6 +186,15 @@ class User:
     # name is username
     name = ast.String("user.name")
 
+    # policies is a list of access policies assigned to the user
+    polices = ast.StringList("user.policies")
```
[the remark this answers is not in the document]
Comment (xacrimon) [reply predicate/solver/teleport.py:190]: They're provided here as a data point to the predicate. They're not for export purposes. My idea for policy_map is to use it to resolve the policies that should apply to a user at the last moment (read: at the point doing the RBAC check). That results in a list of policies we evaluate the predicates of and we can also provide them here.

My reasoning for including this data was to make it easier to work on groups of people within predicates.

[the remark this answers is not in the document]
Comment (vitorenesduarte) [reply predicate/solver/teleport.py:192]: Should we add these 3 when only they're strictly needed?
ATM is not super clear to me if/why they are.

**Comment by vitorenesduarte** [comment]
Are `where` conditions still supported resource rules?
I noticed them in [RFD 44](https://github.com/gravitational/teleport/blob/master/rfd/0044-session-where-condition.md) but they don't seem to be referenced in [the docs](https://goteleport.com/docs/access-controls/reference/).

**Comment by xacrimon** [comment]
> the docs

They are indeed supported and an example can be found here https://goteleport.com/docs/access-controls/reference/#rbac-for-sessions

File: predicate/solver/teleport.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xacrimon) [reply predicate/solver/teleport.py:192]: The main idea was to give all data we have when sensible to increase the flexibility of the system. I don't have any more immediate cases than the one I listed, though. I did end up removing `ssh_logins` as I feel that doesn't mesh too well with the rest of the system.

File: predicate/solver/teleport.py [hunk 3]
```diff
@@ -60,12 +60,28 @@ def collect_like(self, other: ast.Predicate):
...
 
+@scoped
+class Resource(ast.Predicate):
+    """
+    Resource defines read/list/write access to a resource.
+    """
+
+    namespace = ast.String("resource.namespace")
+    kind = ast.String("resource.kind")
+    labels = ast.String("resource.labels")
```
[the remark this answers is not in the document]
Comment (xacrimon) [reply predicate/solver/teleport.py:71]: Including Node doesn't make a lot of sense to me. We don't have data to populate all it's fields at this point. I fixed it and changed it to a stringmap and did some testing.

[the remark this answers is not in the document]
Comment (xacrimon) [reply predicate/solver/teleport.py:71]: Keeping it resource-generic makes more sense IMO as it would also apply to things like k8s resources.

**Review by vitorenesduarte — APPROVED** [review]

**Comment by xacrimon** [comment]
@klizhentas want to give this another look?

File: predicate/examples/list_nodes.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+from solver.teleport import Node, Policy, Resource, Rules
+
+
+class Teleport:
+    p = Policy(
+        name="list_nodes",
+        loud=False,
+        allow=Rules(
+            Resource(
+                (Resource.namespace == "default")
```
Comment (klizhentas) [inline predicate/examples/list_nodes.py:11]: @xacrimon lets remove namespaces for now from Predicate, they are not used, there is only one namespace in all resources today - default one.

File: predicate/solver/teleport.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (klizhentas) [reply predicate/solver/teleport.py:192]: > They're provided here as a data point to the predicate. 

Still, not clear to me what that means, can you elaborate? I understand policy mapping idea that applies when users log in. We can extend policy mapping to apply at different point of interaction between users and resources.

This one is not clear to me, it seems like user is attached policies, how would that interact with policy mapping?

**Review by klizhentas — CHANGES_REQUESTED** [review]

File: predicate/solver/teleport.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xacrimon) [reply predicate/solver/teleport.py:192]: To figure out what policies we need to check, we run the policy mapping code before evaluating predicates to produce a list of policies to apply. We then provide this list as a predefined variable the predicate can check against if it wants to.

**Comment by xacrimon** [comment]
@klizhentas Could you please take another look at this when you have a moment to move it forward? Thanks!

**Review by klizhentas — APPROVED** [review]

**Comment by klizhentas** [comment]
@xacrimon there is one failing test, please fix it and we can merge

**Pull request merged**